server/core/src/main/java/dev/slimevr/tracking/processor/Constraint.kt

```diff
@@ -0,0 +1,183 @@
+package dev.slimevr.tracking.processor
+
+import io.github.axisangles.ktmath.Quaternion
+import io.github.axisangles.ktmath.Vector3
+import kotlin.math.*
+
+/**
+ * Represents a function that applies a rotational constraint.
+ */
+typealias ConstraintFunction = (inputRotation: Quaternion, thisBone: Bone, limit1: Float, limit2: Float, limit3: Float) -> Quaternion
+
+/**
+ * Represents the rotational limits of a Bone relative to its parent,
+ * twist and swing are the max and min when constraintType is a hinge.
+ */
+class Constraint(
+	val constraintType: ConstraintType,
+	twist: Float = 0.0f,
+	swing: Float = 0.0f,
+	allowedDeviation: Float = 0f,
+	maxDeviationFromTracker: Float = 15f,
+) {
+	private val constraintFunction = constraintTypeToFunc(constraintType)
+	private val twistRad = Math.toRadians(twist.toDouble()).toFloat()
+	private val swingRad = Math.toRadians(swing.toDouble()).toFloat()
+	private val allowedDeviationRad = Math.toRadians(allowedDeviation.toDouble()).toFloat()
+	private val maxDeviationFromTrackerRad = Math.toRadians(maxDeviationFromTracker.toDouble()).toFloat()
+	var hasTrackerRotation = false
+
+	/**
+	 * If false solve with minimal movement applied to this link
+	 */
+	var allowModifications = true
+
+	var initialRotation = Quaternion.IDENTITY
+
+	/**
+	 * Apply rotational constraints and if applicable force the rotation
+	 * to be unchanged unless it violates the constraints
+	 */
+	fun applyConstraint(rotation: Quaternion, thisBone: Bone): Quaternion = constraintFunction(rotation, thisBone, swingRad, twistRad, allowedDeviationRad).unit()
+
+	/**
+	 * Force the given rotation to be within allowedDeviation degrees away from
+	 * initialRotation on both the twist and swing axis
+	 */
+	fun constrainToInitialRotation(rotation: Quaternion): Quaternion {
+		val rotationLocal = rotation * initialRotation.inv()
+		var (swingQ, twistQ) = decompose(rotationLocal, Vector3.NEG_Y)
+		swingQ = constrain(swingQ, maxDeviationFromTrackerRad)
+		twistQ = constrain(twistQ, maxDeviationFromTrackerRad)
+		return initialRotation * (swingQ * twistQ)
+	}
+
+	companion object {
+		enum class ConstraintType {
+			TWIST_SWING,
+			HINGE,
+			LOOSE_HINGE,
+			COMPLETE,
+		}
+
+		private fun constraintTypeToFunc(type: ConstraintType) =
+			when (type) {
+				ConstraintType.COMPLETE -> completeConstraint
+				ConstraintType.TWIST_SWING -> twistSwingConstraint
+				ConstraintType.HINGE -> hingeConstraint
+				ConstraintType.LOOSE_HINGE -> looseHingeConstraint
+			}
+
+		private fun decompose(
+			rotation: Quaternion,
+			twistAxis: Vector3,
+		): Pair<Quaternion, Quaternion> {
+			val projection = rotation.project(twistAxis)
+			val twist = Quaternion(sqrt(1.0f - projection.xyz.lenSq()) * if (rotation.w >= 0f) 1f else -1f, projection.xyz).unit()
```

Comment:
I can't think about this right now, but this feels sus doing `1f - xyx.lenSq()`, can I get an explanation on why this logic is correct? There are also other such uses like this within the code.

Reply:
Yes, first of all, it's only correct when the quaternion is unitized, so in this case it is very much sus because the input is not unitized (I fixed it now but it was broken good catch). But the idea is that when you have a unitized quaternion the scalar part will be equal to the square root of one minus the vector parts length squared because the total length of the quaternion is one. You have to reconstruct the w component in this way because the vector part's length may have been changed when it was projected onto an axis and if the w component does not bring the length of the quaternion to 1 then the rotation will not be correct. Using just the projected rotation does work but there are some inputs that cause an invalid quaternion to be created which I want to avoid at all costs. I hope that explanation was understandable!